### PR TITLE
Add checkstyle.xml, idea-style guide (For importing in IDE) and enabl…

### PR DESCRIPTION
**Pull request by dhaval24**
Add CheckStyle rules.

Related #490 #694 

@littleaj let's review the rules here so we can make a bulk change in the whole SDK.

### CODE REVIEW

File: config/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        </module>
+        <module name="EmptyCatchBlock">
+            <property name="exceptionVariableName" value="expected"/>
+        </module>
+        <module name="CommentsIndentation"/>
+        <module name="SuppressWarningsHolder"/>
+        <module name="SuppressionCommentFilter"/>
+    </module>
+    <module name="SuppressWarningsFilter"/>
+</module>
```
Comment (reyang) [inline config/checkstyle/checkstyle.xml]: Do we want an empty line before EOF?

File: config/checkstyle/checkstyle.xml [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        <module name="RedundantImport"/>
+        <module name="OneTopLevelClass"/>
+        <module name="NoLineWrap"/>
+        <module name="EmptyBlock">
+            <property name="option" value="TEXT"/>
+            <property name="tokens"
+                      value="LITERAL_TRY, LITERAL_FINALLY, LITERAL_IF, LITERAL_ELSE, LITERAL_SWITCH"/>
+        </module>
+        <module name="NeedBraces"/>
+        <module name="LeftCurly"/>
```
Comment (littleaj) [inline config/checkstyle/checkstyle.xml]: I would be OK with this ("new line on wrap"):
```suggestion
        <module name="LeftCurly">
                <property name="option" value="nlow" />
        </module>
```
but I'm ok with `eol`, too. It's up to you.

File: config/checkstyle/checkstyle.xml [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+            <property name="ignoreFinal" value="false"/>
+            <property name="allowedAbbreviationLength" value="1"/>
+        </module>
+        <module name="OverloadMethodsDeclarationOrder"/>
+        <!-- <!-\- Many unit tests define all variables at the start of the method. -\-> -->
+        <!-- <module name="VariableDeclarationUsageDistance"/> -->
+        <module name="CustomImportOrder">
+            <property name="sortImportsInGroupAlphabetically" value="true"/>
+            <property name="separateLineBetweenGroups" value="true"/>
+            <property name="customImportOrderRules" value="STATIC###THIRD_PARTY_PACKAGE"/>
```
Comment (littleaj) [inline config/checkstyle/checkstyle.xml]: How do you feel about this: 
```suggestion
            <property name="customImportOrderRules" value="STANDARD_JAVA_PACKAGE###THIRD_PARTY_PACKAGE###STATIC"/>
```

File: config/checkstyle/checkstyle.xml [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+            <property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9_]*$"/>
+            <message key="name.invalidPattern"
+                     value="Method name ''{0}'' must match pattern ''{1}''."/>
+        </module>
+        <module name="SingleLineJavadoc">
+            <!-- <!-\- Wrong interpretation of the style guide; -\-> -->
+            <!-- <property name="ignoreInlineTags" value="false"/ -->
+        </module>
+        <module name="EmptyCatchBlock">
+            <property name="exceptionVariableName" value="expected"/>
```
Comment (littleaj) [inline config/checkstyle/checkstyle.xml]:
```suggestion
            <property name="exceptionVariableName" value="expected|ignore|ignoreme|dontcare" />
```

File: config/checkstyle/checkstyle.xml [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        </module>
+        <module name="CommentsIndentation"/>
```
Comment (littleaj) [inline config/checkstyle/checkstyle.xml]: Some additions I'd like to see
```XML
        <module name="HiddenField" /><!-- warn when fields are being masked -->
        <module name="MaigcNumber" /><!-- don't use magic numbers -->
        <module name="Header">
                <property name="headerFile" value="java.header" /> <!-- this file will contain the copyright notice -->
                <property name="fileExtensions" value="java" />
        </module>
```

File: config/checkstyle/intellij-java-google-style.xml [hunk 6]
```diff
@@ -0,0 +1,598 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (dhaval24) [reply config/checkstyle/intellij-java-google-style.xml]: This file basically helps you to import the style guide in intellij and then you can use short cuts to format your code.

File: config/checkstyle/checkstyle.xml (hunk 2 above)
Comment (dhaval24) [reply config/checkstyle/checkstyle.xml]: I would like eol. In java projects (atleast what I have seen) people tend to prefer eol bracing strategy.

File: config/checkstyle/checkstyle.xml (hunk 3 above)
Comment (dhaval24) [reply config/checkstyle/checkstyle.xml]: This one looks better. I will update this rule.

File: config/checkstyle/checkstyle.xml (hunk 4 above)
Comment (dhaval24) [reply config/checkstyle/checkstyle.xml]: what's the rationale behind this. Do we want to use names like `ignore|ignoreme|dontcare` for exceptions. Specifically the last 2 names don't look good to me on public facing project.

File: config/checkstyle/checkstyle.xml (hunk 5 above)
Comment (dhaval24) [reply config/checkstyle/checkstyle.xml]: Magic number is good to not have. However, I have a feeling that this kind of coding practice might be prevelant in our project and cannot be undone automatically. We can turn on the rule and update incrementally.
For Header rule, I don't have any problems.

File: config/checkstyle/checkstyle.xml (hunk 4 above)
Comment (littleaj) [reply config/checkstyle/checkstyle.xml]: haha...fair enough. I just wanted to be more relaxed. I'm not too worried about this one. we can leave it as-is.

File: config/checkstyle/checkstyle.xml [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+        <module name="AvoidEscapedUnicodeCharacters">
+            <property name="allowEscapesForControlCharacters" value="true"/>
+            <property name="allowByTailComment" value="true"/>
+            <property name="allowNonPrintableEscapes" value="true"/>
+        </module>
+        <module name="LineLength">
+            <property name="max" value="120"/>
+            <property name="ignorePattern" value="^package.*|^import.*|a href|href|http://|https://|ftp://"/>
+        </module>
+        <module name="AvoidStarImport"/>
```
[the remark this answers is not in the document]
Comment (dhaval24) [reply config/checkstyle/checkstyle.xml]: Do you really feel we should have star import? Is this a common practice in OSS? If yes, I am fine to add this. Let me know what you think

File: config/checkstyle/checkstyle.xml (hunk 1 above)
Comment (reyang) [reply config/checkstyle/checkstyle.xml]: General practice is to add empty line before EOF.

**Comment by dhaval24** [comment]
@littleaj can we conclude on this PR today? Let me know if there are additional things in your mind.

File: config/checkstyle/checkstyle.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (dhaval24) [reply config/checkstyle/checkstyle.xml]: Updated.

**Pull request merged**